### PR TITLE
Add separate publisher for JSON output

### PR DESCRIPTION
**Pull request by cescgina**
Producing JSON output for multiple environments requires accumulating
the environments string representation and print it all at once, after
all queries have finished. On the other hand, when printing to screen,
we want to show the output as soon as each environment is queried,
so this change introduces two Publisher classes, one for each use
case.

### CODE REVIEW

File: cibyl/cli/main.py [hunk 1]
```diff
@@ -122,10 +122,9 @@ def main() -> None:
         orchestrator.validate_environments()
         features = orchestrator.load_features()
         orchestrator.query_and_publish(
-            output_path=arguments["output_file_path"],
-            output_style=arguments["output_style"],
-            features=features
-        )
+                output_path=arguments["output_file_path"],
+                output_style=arguments["output_style"],
+                features=features)
```
Comment (jsanemet) [inline cibyl/cli/main.py:127]: Unnecessary change, but well, it does not really matter.

File: cibyl/orchestrator.py [hunk 2]
```diff
@@ -68,7 +68,8 @@ def __init__(self, environments: list = None):
         """Orchestrator constructor method"""
         self.parser = Parser()
         self.config = AppConfig()
-        self.publisher = Publisher()
+        # publisher will be initialized when needed
+        self.publisher = None
```
Comment (jsanemet) [inline cibyl/orchestrator.py:72]: You may remove this variable, the publisher is only used in one function.

File: cibyl/orchestrator.py [hunk 3]
```diff
@@ -386,27 +395,27 @@ def query():
...
+                style=output_style,
+                query=query_type,
+                verbosity=self.parser.app_args.get('verbosity', 0),
+                output_file=file)
 
         for env in self.environments:
             query()
-            publish()
+            self.publisher.publish(environment=env)
+        self.publisher.finish_publishing()
```
Comment (jsanemet) [inline cibyl/orchestrator.py:421]: Overall, I think that the concept behind the publisher class is kind of outdated. In the end, it does not do much more than appending output from the printers. I think that all of this can be simplified with the IndentedTextBuilder tool and a couple of if statements. For example, these could be swapped for:
```
output = Indented...()

for ...:
    output.add(printer.print_env(env), 0)

output.build()
```

**Review by jsanemet — CHANGES_REQUESTED** [review]
Really great work. I have left a couple of suggestions, see what you think.

The only real request that I have got, is that instead of an array, return the answers as a section 'environments':
```
{
  'env...': [
    { 'name': '...' }
  ]
}
```
That way it is clear what those are and it is easier to extend later on.

File: cibyl/cli/main.py (hunk 1 above)
Comment (cescgina) [reply cibyl/cli/main.py:127]: While trying things out I removed the arguments and added them back, while editing I must have moved the parenthesis without realising. I'm surprise it marks the three lines as changed though.

File: cibyl/orchestrator.py (hunk 2 above)
Comment (cescgina) [reply cibyl/orchestrator.py:72]: True, I'll remove it.

File: cibyl/orchestrator.py (hunk 3 above)
Comment (cescgina) [reply cibyl/orchestrator.py]: I have tried your suggestion and you're right, it can be done (I have uploaded a branch with the change to my fork, [link](https://github.com/cescgina/cibyl/blob/d3fb2e6103df41c21d6aa9e6dedd34e8b7b4f5fe/cibyl/orchestrator.py#L413)). It's somewhat simpler, but I'm not sure if it would be simpler in the long run. If for example in the future we want to support yaml files, we will need to add another branch to the if, and then I'm not sure we're actually better off.

**Comment by cescgina** [comment]
Thanks for the comments, @jsanemet. I have updated the output for the json format.

File: cibyl/orchestrator.py (hunk 3 above)
Comment (jsanemet) [reply cibyl/orchestrator.py]: Ok, let's go with this one then.

**Review by jsanemet — APPROVED** [review]
